helix-core/src/words.rs

```diff
@@ -1,35 +1,19 @@
-use crate::movement::{categorize, is_punctuation, is_word, skip_over_prev};
+use crate::movement::{
+    backwards_skip_while, categorize, is_end_of_line, is_punctuation, is_word, Category,
+    SliceIndexHelpers,
+};
 use ropey::RopeSlice;
 
 #[must_use]
-pub fn nth_prev_word_boundary(slice: RopeSlice, mut char_idx: usize, count: usize) -> usize {
-    let mut with_end = false;
-
-    for _ in 0..count {
-        if char_idx == 0 {
-            break;
-        }
-
-        // return if not skip while?
-        skip_over_prev(slice, &mut char_idx, |ch| ch == '\n');
-
-        with_end = skip_over_prev(slice, &mut char_idx, char::is_whitespace);
-
-        // refetch
-        let ch = slice.char(char_idx);
-
-        if is_word(ch) {
-            with_end = skip_over_prev(slice, &mut char_idx, is_word);
-        } else if is_punctuation(ch) {
-            with_end = skip_over_prev(slice, &mut char_idx, is_punctuation);
-        }
-    }
-
-    if with_end || char_idx == 0 {
-        char_idx
-    } else {
-        char_idx + 1
-    }
+pub fn nth_prev_word_boundary(slice: RopeSlice, index: usize, count: usize) -> usize {
+    (0..count).fold(index, |mut index, _| {
+        index = backwards_skip_while(slice, index, is_end_of_line).unwrap_or(index);
+        index = backwards_skip_while(slice, index, char::is_whitespace).unwrap_or(index);
+        let category = index.category(slice).unwrap_or(Category::Unknown);
+        backwards_skip_while(slice, index, |c| categorize(c) == category)
+            .map(|i| i + 1)
+            .unwrap_or(0)
+    })
```

Comment:
We have this pattern quite often. I wonder if we can shorten it to something like `xxx.repeat(count)`?